File: matlab/+okf/fetch.m
function f = fetch(source, subdir, branch)
%FETCH Materialize a bundle from a dir, local tar/zip archive, or git URL.
%   F = OKF.FETCH(SOURCE[, SUBDIR[, BRANCH]]) returns a struct with fields
%   dir, kind, and cleanup (a function handle removing the temp dir; call it
%   when done -- okf.ingest does this automatically).
%
%   Archives extract via the untar/unzip builtins, which cannot list members
%   before extraction; instead extraction goes into a FRESH temp dir and
%   every returned path is verified to sit inside it afterwards (containment
%   check). Remote http(s) archive download is not supported (use the
%   R/Python binding).
if nargin < 2
    subdir = '';
end
if nargin < 3
    branch = '';
end
f = struct('dir', '', 'kind', '', 'cleanup', @() []);
if exist(source, 'dir') == 7
    old = cd(source);
    f.dir = pwd;
    cd(old);
    f.kind = 'dir';
    return;
end
kind = source_kind(source);
if ~isempty(regexp(source, '^https?://', 'once')) && ~strcmp(kind, 'git')
    error('okf:fetch', ['remote archive download is not supported by the MATLAB ' ...
                        'binding; fetch the archive locally first (git URLs are supported)']);
end
tmp = tempname();
mkdir(tmp);
f.cleanup = @() rm_rf(tmp);
try
    if strcmp(kind, 'git')
        repo = fullfile(tmp, 'repo');
        cmd = 'git clone --depth 1 ';
        if ~isempty(branch)
            cmd = [cmd '--branch "' branch '" '];
        end
        cmd = [cmd '"' source '" "' repo '"'];
        [rc, ~] = system(cmd);
        if rc ~= 0
            error('okf:fetch', 'git clone failed (is git installed?): %s', source);
        end
        base = repo;
    else
        ex = fullfile(tmp, 'x');
        mkdir(ex);
        if strcmp(kind, 'zip')
            extracted = unzip(source, ex);
        else
            extracted = untar(source, ex);
        end
        assert_contained(extracted, ex);
        base = ex;
    end
    f.dir = bundle_root(base, subdir);
    f.kind = kind;
catch err
    f.cleanup();
    rethrow(err);
end
end

function kind = source_kind(source)
s = regexprep(source, '[?#].*$', '');
ls = lower(s);
if okf.internal.ends_with(ls, '.zip')
    kind = 'zip';
elseif okf.internal.ends_with(ls, '.tar.gz') || okf.internal.ends_with(ls, '.tgz') || ...
       okf.internal.ends_with(ls, '.tar') || okf.internal.ends_with(ls, '.tar.bz2')
    kind = 'tar';
elseif okf.internal.ends_with(s, '.git') || strncmp(source, 'git@', 4) || ...
       ~isempty(regexp(s, '^https?://(www\.)?(github|gitlab|bitbucket)\.', 'once'))
    kind = 'git';
else
    error('okf:fetch', ...
          'cannot determine source kind (expected a dir, git URL, or tar/zip): %s', source);
end
end

function assert_contained(extracted, ex)
%ASSERT_CONTAINED Post-extraction traversal check: every extracted path must
%   sit inside the fresh extraction dir (untar/unzip give no pre-listing).
%   MATLAB untar returns absolute paths; Octave returns member-relative ones
%   -- reduce both to a path relative to EX and reject escapes.
old = cd(ex);
ex_abs = pwd;
cd(old);
prefix = [ex_abs filesep];
for i = 1:numel(extracted)
    p = extracted{i};
    is_abs = (numel(p) >= 2 && p(2) == ':') || (~isempty(p) && (p(1) == '/' || p(1) == '\'));
    if is_abs
        if strcmp(p, ex_abs) || strncmp(p, prefix, numel(prefix))
            rel = p(min(numel(prefix) + 1, numel(p) + 1):end);
        else
            rm_rf(ex_abs);
            error('okf:fetch', 'archive member escaped target dir (path traversal): %s', p);
        end
    else
        rel = p;
    end
    segs = strsplit(strrep(rel, '\', '/'), '/');
    if any(strcmp(segs, '..'))
        rm_rf(ex_abs);
        error('okf:fetch', 'archive member escaped target dir (path traversal): %s', p);
    end
end
end

function root = bundle_root(base, subdir)
if ~isempty(subdir)
    root = fullfile(base, subdir);
    return;
end
cur = base;
for hop = 1:6
    entries = dir(cur);
    has_md = false;
    dirs = {};
    for i = 1:numel(entries)
        name = entries(i).name;
        if strcmp(name, '.') || strcmp(name, '..') || name(1) == '.'
            continue;
        end
        if okf.internal.ends_with(lower(name), '.md')
            has_md = true;
        end
        if entries(i).isdir
            dirs{end + 1} = fullfile(cur, name); %#ok<AGROW>
        end
    end
    if ~has_md && numel(dirs) == 1
        cur = dirs{1};
    else
        break;
    end
end
root = cur;
end

function rm_rf(d)
if exist(d, 'dir') == 7
    try
        rmdir(d, 's');
    catch
    end
end
end

File: rust/okf-ingest/src/fetch.rs
//! Bundle materialization — mirrors `py/okf/okf.py::fetch` (`_source_kind`,
//! `_assert_safe_members`, `_bundle_root`), offline scope: local dir, local
//! tar/zip archive, or git URL via the `git` subprocess. Remote http(s)
//! archive download is not supported in this binding (document; the R and
//! Python bindings download).

use std::collections::BTreeSet;
use std::fs::File;
use std::path::{Component, Path, PathBuf};
use std::process::Command;

use flate2::read::GzDecoder;
use tempfile::TempDir;

use crate::model::{OkfError, SourceKind};

/// A materialized bundle directory. Holding the value keeps any temp dir
/// alive; it is removed on drop (the Python binding's `cleanup()` closure).
pub struct Fetched {
    pub dir: PathBuf,
    pub kind: SourceKind,
    _tmp: Option<TempDir>,
}

fn strip_query(source: &str) -> &str {
    let end = source.find(['?', '#']).unwrap_or(source.len());
    &source[..end]
}

fn source_kind(source: &str) -> Result<SourceKind, OkfError> {
    let s = strip_query(source);
    let ls = s.to_lowercase();
    if ls.ends_with(".zip") {
        return Ok(SourceKind::Zip);
    }
    if ls.ends_with(".tar.gz")
        || ls.ends_with(".tgz")
        || ls.ends_with(".tar")
        || ls.ends_with(".tar.bz2")
    {
        return Ok(SourceKind::Tar);
    }
    let host_is_forge = regex::Regex::new(r"^https?://(www\.)?(github|gitlab|bitbucket)\.")
        .unwrap()
        .is_match(s);
    if s.ends_with(".git") || source.starts_with("git@") || host_is_forge {
        return Ok(SourceKind::Git);
    }
    Err(OkfError::msg(format!(
        "cannot determine source kind (expected a dir, git URL, or tar/zip): {source}"
    )))
}

/// Reject archive members that would extract outside the target directory
/// (path traversal / zip-slip), before extracting anything.
fn assert_safe_member(name: &str) -> Result<(), OkfError> {
    let p = Path::new(name);
    for comp in p.components() {
        match comp {
            Component::ParentDir | Component::RootDir | Component::Prefix(_) => {
                return Err(OkfError::msg(format!(
                    "archive member escapes target dir (path traversal): {name:?}"
                )));
            }
            _ => {}
        }
    }
    Ok(())
}

/// Descend into a single wrapping directory (up to 6 levels) to find the
/// bundle root — archives usually wrap their content in one top-level dir.
fn bundle_root(base: &Path, subdir: Option<&str>) -> Result<PathBuf, OkfError> {
    if let Some(sd) = subdir {
        return Ok(base.join(sd));
    }
    let mut cur = base.to_path_buf();
    for _ in 0..6 {
        let mut has_md = false;
        let mut dirs: Vec<PathBuf> = Vec::new();
        for entry in std::fs::read_dir(&cur)? {
            let entry = entry?;
            let name = entry.file_name().to_string_lossy().to_string();
            if name.starts_with('.') {
                continue;
            }
            if name.to_lowercase().ends_with(".md") {
                has_md = true;
            }
            if entry.path().is_dir() {
                dirs.push(entry.path());
            }
        }
        if !has_md && dirs.len() == 1 {
            cur = dirs.remove(0);
        } else {
            break;
        }
    }
    Ok(cur)
}

fn extract_tar(local: &Path, dst: &Path) -> Result<(), OkfError> {
    let ls = local.to_string_lossy().to_lowercase();
    if ls.ends_with(".tar.bz2") {
        return Err(OkfError::msg(
            ".tar.bz2 is not supported by the Rust binding (use .tar.gz)",
        ));
    }
    let gz = ls.ends_with(".tar.gz") || ls.ends_with(".tgz");
    // Pass 1: safety check on all member names.
    let names: BTreeSet<String> = {
        let f = File::open(local)?;
        let mut entries_names = BTreeSet::new();
        if gz {
            let mut a = tar::Archive::new(GzDecoder::new(f));
            for e in a.entries()? {
                let e = e?;
                entries_names.insert(e.path()?.to_string_lossy().to_string());
            }
        } else {
            let mut a = tar::Archive::new(f);
            for e in a.entries()? {
                let e = e?;
                entries_names.insert(e.path()?.to_string_lossy().to_string());
            }
        }
        entries_names
    };
    for n in &names {
        assert_safe_member(n)?;
    }
    // Pass 2: extract.
    let f = File::open(local)?;
    if gz {
        tar::Archive::new(GzDecoder::new(f))
            .unpack(dst)
            .map_err(|e| OkfError::msg(format!("tar extract failed: {e}")))?;
    } else {
        tar::Archive::new(f)
            .unpack(dst)
            .map_err(|e| OkfError::msg(format!("tar extract failed: {e}")))?;
    }
    Ok(())
}

fn extract_zip(local: &Path, dst: &Path) -> Result<(), OkfError> {
    let f = File::open(local)?;
    let mut archive =
        zip::ZipArchive::new(f).map_err(|e| OkfError::msg(format!("zip open failed: {e}")))?;
    let names: Vec<String> = archive.file_names().map(str::to_string).collect();
    for n in &names {
        assert_safe_member(n)?;
    }
    archive
        .extract(dst)
        .map_err(|e| OkfError::msg(format!("zip extract failed: {e}")))?;
    Ok(())
}

/// Materialize a bundle from a dir, git URL, or local tar/zip archive.
pub fn fetch(
    source: &str,
    subdir: Option<&str>,
    branch: Option<&str>,
) -> Result<Fetched, OkfError> {
    let src_path = Path::new(source);
    if src_path.is_dir() {
        return Ok(Fetched {
            dir: std::fs::canonicalize(src_path)?,
            kind: SourceKind::Dir,
            _tmp: None,
        });
    }
    let kind = source_kind(source)?;
    if regex::Regex::new(r"^https?://").unwrap().is_match(source) && kind != SourceKind::Git {
        return Err(OkfError::msg(
            "remote archive download is not supported by the Rust binding; \
             fetch the archive locally first (git URLs are supported)",
        ));
    }
    let tmp = TempDir::with_prefix("okf_")?;
    let base = if kind == SourceKind::Git {
        let repo = tmp.path().join("repo");
        let mut cmd = Command::new("git");
        cmd.args(["clone", "--depth", "1"]);
        if let Some(b) = branch {
            cmd.args(["--branch", b]);
        }
        cmd.arg(source).arg(&repo);
        let status = cmd
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .status()
            .map_err(|e| OkfError::msg(format!("git clone failed (is git installed?): {e}")))?;
        if !status.success() {
            return Err(OkfError::msg(format!(
                "git clone failed (is git installed?): {source}"
            )));
        }
        repo
    } else {
        let ex = tmp.path().join("x");
        std::fs::create_dir_all(&ex)?;
        match kind {
            SourceKind::Zip => extract_zip(src_path, &ex)?,
            _ => extract_tar(src_path, &ex)?,
        }
        ex
    };
    let dir = bundle_root(&base, subdir)?;
    Ok(Fetched {
        dir,
        kind,
        _tmp: Some(tmp),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn kind_detection() {
        assert_eq!(source_kind("x/b.zip").unwrap(), SourceKind::Zip);
        assert_eq!(source_kind("x/b.tar.gz").unwrap(), SourceKind::Tar);
        assert_eq!(source_kind("x/b.tgz").unwrap(), SourceKind::Tar);
        assert_eq!(
            source_kind("git@github.com:a/b.git").unwrap(),
            SourceKind::Git
        );
        assert_eq!(
            source_kind("https://github.com/a/b").unwrap(),
            SourceKind::Git
        );
        assert!(source_kind("what.is.this").is_err());
    }

    #[test]
    fn traversal_guard() {
        assert!(assert_safe_member("store/index.md").is_ok());
        assert!(assert_safe_member("../evil.md").is_err());
        assert!(assert_safe_member("/abs.md").is_err());
        assert!(assert_safe_member("a/../../evil.md").is_err());
    }
}
